### PR TITLE
refactor: add tr_torrent::Error helper class

### PR DESCRIPTION
**Pull request by ckerr**
Another incremental PR to reduce `tr_torrent`  sprawl by dividing its code into more self-contained chunks.

This replaces three error fields (`error`, `error_string`, and `error_announce_url`) with a single `error` field that is an instance of a new helper class to improve semantics, e.g. client code can call `error().set_tracker_warning()` instead of manually updating all the fields.

### CODE REVIEW

File: libtransmission/inout.cc [hunk 1]
```diff
@@ -218,11 +218,11 @@ int readOrWritePiece(tr_torrent* tor, IoMode io_mode, tr_block_info::Location lo
         tr_error* error = nullptr;
         readOrWriteBytes(tor->session, tor, io_mode, file_index, file_offset, buf, bytes_this_pass, &error);
 
-        if (error != nullptr)
+        if (error != nullptr) // if IO failed, set torrent's error if not already set
         {
-            if (io_mode == IoMode::Write && tor->error != TR_STAT_LOCAL_ERROR)
+            if (io_mode == IoMode::Write && std::empty(tor->error()))
```
Comment (tearfur) [inline libtransmission/inout.cc:223]: Is this change intended?

File: libtransmission/torrent.cc [hunk 2]
```diff
@@ -2282,9 +2311,9 @@ void tr_torrent::set_download_dir(std::string_view path, bool is_new_torrent)
             recheck_completeness();
         }
     }
-    else if (error == TR_STAT_LOCAL_ERROR && !setLocalErrorIfFilesDisappeared(this))
+    else if (!std::empty(error_) && !setLocalErrorIfFilesDisappeared(this))
```
Comment (tearfur) [inline libtransmission/torrent.cc:2314]: Is it better to have the "files disappeared" local error overwrite the existing tracker warning/error?

If no, then this will revert the behaviour:

```suggestion
    else if (error_.error_type() != TR_STAT_LOCAL_ERROR && !setLocalErrorIfFilesDisappeared(this))
```

File: libtransmission/inout.cc (hunk 1 above)
Comment (ckerr) [reply libtransmission/inout.cc]: thanks. reverted in 599a426a1f383ed820c72a54e1ebbc9b13084200

File: libtransmission/torrent.cc (hunk 2 above)
Comment (ckerr) [reply libtransmission/torrent.cc]: restored it to the previous code in e9dd5001122679ad7a8e463c8c481e486b140380

**Review by tearfur — APPROVED** [review]
